src/middlewared/middlewared/plugins/vm.py

```diff
@@ -1,487 +1,651 @@
-            ]
+        return domain
+
+    def os_xml(self):
+        # TODO: Ensure bios works
+        os_list = []
+        children = [create_element('type', attribute_dict={'text': 'hvm'})]
+        if self.vm_data['bootloader'] in ('UEFI', 'UEFI_CSM'):
+            children.append(
+                create_element(
+                    'loader', attribute_dict={
+                        'text': '/usr/local/share/uefi-firmware/BHYVE_UEFI'
+                        f'{"_CSM" if self.vm_data["bootloader"] == "UEFI_CSM" else ""}.fd'
+                    }, readonly='yes', type='pflash',
+                ),
+            )
 
-        if self.vm['time'] == 'UTC':
-            args += ['-u']
+        if self.vm_data['bootloader'] == 'GRUB':
+            # Following is keeping compatibility with old code where we supported rancher/docker
+            # We can have two cases where the user has the path set pointing to his/her grub file or where
+            # the data inside grubconfig is saved and we write out a new file for it
+            # In either of these cases we require device map file to be written
 
-        nid = Nid(3)
-        grub_devices = []
-        block_devices = {
-            'ahci': [],
-            'virtio-blk': [],
-        }
-        for device in sorted(self.vm['devices'], key=lambda x: (x['order'], x['id'])):
-            if device['dtype'] in ('CDROM', 'DISK', 'RAW'):
+            device_map_data = '\n'.join(
+                f'(hd{i}) {d.data["attributes"]["path"]}' for i, d in enumerate(filter(
+                    lambda d: isinstance(d, RAW) and d.data['attributes'].get('boot'), self.devices
+                ))
+            )
+            # It will be ensured while starting that this VM is capable of starting i.e has boot devices
+
+            os_list.append(create_element(
+                'bootloader', attribute_dict={'text': '/usr/local/sbin/grub-bhyve'}
+            ))
 
-                # Get grub devices to be used in grub-bhyve
-                if device['dtype'] == 'RAW' and device['attributes'].get('boot'):
-                    grub_devices.append(device['attributes']['path'])
+            device_map_dir = os.path.join('/tmp/grub', self.libvirt_domain_name)
+            device_map_file = os.path.join(device_map_dir, 'devices_map')
+            os.makedirs(device_map_dir, exist_ok=True)
 
-                disk_sector_size = int(device['attributes'].get('sectorsize') or 0)
-                if disk_sector_size > 0:
-                    sectorsize_args = ',sectorsize=' + str(disk_sector_size)
+            with open(device_map_file, 'w') as f:
+                f.write(device_map_data)
+
+            if self.vm_data['grubconfig']:
+                grub_config = self.vm_data['grubconfig'].strip()
+                if grub_config.startswith('/mnt') and os.path.exists(grub_config):
```

Comment:
What does `startswith('/mnt')` check help to do?

Reply:
I think it ensures that we don't have this somewhere else in the system. These bits have been copied from legacy logic to support old behavior.